content/concepts/hole-punching.md

```diff
@@ -0,0 +1,193 @@
+for central coordination servers like STUN and TURN. 
+
+The following sequence diagram illustrates the whole process.
+
+![](https://i.imgur.com/sMGMfGZ.png)
+
+libp2p hole punching can be derived in two phases, a preparation phase and 
+a hole punching phase.
+
+### Phase I: Preparation
+
+1. [AutoNAT](/concepts/nat/#autonat): Determine whether a node is dialable, 
+   as in, discover if a node is behind a NAT or firewall.
+   
+   > This is equivalent to the 
+   > [STUN protocol](https://www.rfc-editor.org/rfc/rfc3489) in ICE.
+
+   ![](https://i.imgur.com/NeufUm7.png)
+   
+   - `B` reaches out to `Other_Peers` (e.g., boot nodes) on the network it 
+     is on and asks each node to dial it on a set of addresses it suspects 
+     could be reachable. 
+   - `Other_Peers`attempt to dial each of `B`'s addresses and report the 
+     outcome back to `B`. 
+   - Based on the reports, `B` can gauge whether it is publicly dialable and 
+     determine if hole punching is needed.
+   
+<!-- to add routing reference when available -->
+1. Routing: Discover the k-closest public Relay nodes using a lookup method 
+   (e.g. IPFS uses Kademlia DHT): `/<RELAY_ADDR>/p2p-circuit/<PEER_ID_B>`
+
+    ![](https://i.imgur.com/cdqmJCo.png)
+    
+    - `Other_Peers` outside `B`'s network can dial `B` indirectly through 
+      a public Relay node. In the case of [IPFS](https://ipfs.tech/), each public 
+      node would serve as a `Relay`. `B` would either perform a lookup on the 
+      [Kademlia DHT](https://github.com/libp2p/specs/blob/master/kad-dht/README.md) 
+      for the closest peers to its Peer ID or choose a subset of the public nodes 
+      it is already connected to.
+   
+2. [Circuit Relay](/concepts/circuit-relay): Connect to and request 
+   reservations with the discovered Relay nodes. A node can advertise itself as being 
+   reachable through a remote Relay node. 
+   
+   > This is equivalent to the 
+   > [TURN protocol](https://datatracker.ietf.org/doc/html/rfc5766) in ICE.
+
+    ![](https://i.imgur.com/kqRnQUV.png)
+
+   - `Relay` can limit the resources used to relay connections (e.g., by the number 
+     of connections, the time, and bytes) via Circuit Relay v2. In the case of IPFS, 
+     this allows every public node in the network to serve as a relay without high 
+     resource consumption.
+   - For each discovered `Relay`, `B`: 
+       - connects to the remote node and requests the Relay node to listen to 
+         connections on its behalf, known as a reservation;
+       - if `Relay` accepts reservation requests, `B` can advertise itself as being 
+         reachable through `Relay`.
+
+### Phase II: Hole punching
+
+1. [Circuit Relay](/concepts/circuit-relay): Establish a secure relay connection 
+   through the public Relay node. The node establishes a direct connection with 
+   the Relay node. It then requests a relayed connection to the other node through 
```

Comment:
_The other node_ requests to the relayed connection. Maybe it makes sense to use A and B here, to make clear which node we’re talking about?